R/chrome.R

```diff
@@ -86,56 +86,65 @@ launch_chrome <- function(path = find_chrome(), args = get_chrome_args()) {
     stop("Invalid path to Chrome")
   }
 
-  p <- process$new(
-    command = path,
-    args = c("--headless", "--remote-debugging-port=0", args),
-    supervise = TRUE,
-    stdout = tempfile("chrome-stdout-", fileext = ".log"),
-    stderr = tempfile("chrome-stderr-", fileext = ".log")
-  )
-
+  with_random_port(function(host, port) {
```

Comment:
I find it a bit difficult to keep track of state when a closures are created inside of functions and passed around like this. I think it would be clearer to pull out the code in this function to a top level function, which makes it easier to reason about what exactly the function takes as inputs, and is also easier to test.

Also, the structure of the current code in this PR obscures the fact that the value returned from the anonymous function (`list(process=p, port=port)`) is also returned by `with_random_port()`, and that `launch_chrome()` is also supposed to return that value. I suggest changing the code to make that more obvious.

Something like this:

```R
launch_chrome_internal <- function(path, args, host, port) {
  # Function implementation here ...
}

launch_chrome <- function(path = find_chrome(), args = get_chrome_args()) {
  res <- with_random_port(function(port) {
    launch_chrome_internal(path, args, "http://127.0.0.1", port)
  })
  
  res
}
```

Also note that `with_random_port()` now takes a function that has only one parameter, `port`.

Reply:
I completely agree. I originally tried to wrap just `processx::run()` but later realized I needed to include the rest of the logic in `launch_chrome()`. I added `launch_chrome_impl()` that uses the pattern you described.